### PR TITLE
Fix #15297: increase chance of unique `episode_id`

### LINKED ISSUE #15297: [rllib] Assert agent_key not in self.agent_collectors
### What is the problem?

After a couple of training iterations, the training job crashes with the following error:

```
 Failure # 1 (occurred at 2021-03-31_11-10-22)
Traceback (most recent call last):
  File "/opt/miniconda/lib/python3.7/site-packages/ray/tune/trial_runner.py", line 586, in _process_trial
    results = self.trial_executor.fetch_result(trial)
  File "/opt/miniconda/lib/python3.7/site-packages/ray/tune/ray_trial_executor.py", line 609, in fetch_result
    result = ray.get(trial_future[0], timeout=DEFAULT_GET_TIMEOUT)
  File "/opt/miniconda/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 47, in wrapper
    return func(*args, **kwargs)
  File "/opt/miniconda/lib/python3.7/site-packages/ray/worker.py", line 1456, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(AssertionError): e[36mray::PPO.train_buffered()e[39m (pid=264, ip=10.1.0.8)
  File "python/ray/_raylet.pyx", line 480, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 432, in ray._raylet.execute_task.function_executor
  File "/opt/miniconda/lib/python3.7/site-packages/ray/tune/trainable.py", line 167, in train_buffered
    result = self.train()
  File "/opt/miniconda/lib/python3.7/site-packages/ray/rllib/agents/trainer.py", line 526, in train
    raise e
  File "/opt/miniconda/lib/python3.7/site-packages/ray/rllib/agents/trainer.py", line 515, in train
    result = Trainable.train(self)
  File "/opt/miniconda/lib/python3.7/site-packages/ray/tune/trainable.py", line 226, in train
    result = self.step()
  File "/opt/miniconda/lib/python3.7/site-packages/ray/rllib/agents/trainer_template.py", line 157, in step
    evaluation_metrics = self._evaluate()
  File "/opt/miniconda/lib/python3.7/site-packages/ray/rllib/agents/trainer.py", line 778, in _evaluate
    for w in self.evaluation_workers.remote_workers()
  File "/opt/miniconda/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 47, in wrapper
    return func(*args, **kwargs)
ray.exceptions.RayTaskError(AssertionError): e[36mray::RolloutWorker.sample()e[39m (pid=375, ip=10.1.0.8)
  File "python/ray/_raylet.pyx", line 480, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 432, in ray._raylet.execute_task.function_executor
  File "/opt/miniconda/lib/python3.7/site-packages/ray/rllib/evaluation/rollout_worker.py", line 662, in sample
    batches = [self.input_reader.next()]
  File "/opt/miniconda/lib/python3.7/site-packages/ray/rllib/evaluation/sampler.py", line 95, in next
    batches = [self.get_data()]
  File "/opt/miniconda/lib/python3.7/site-packages/ray/rllib/evaluation/sampler.py", line 224, in get_data
    item = next(self.rollout_provider)
  File "/opt/miniconda/lib/python3.7/site-packages/ray/rllib/evaluation/sampler.py", line 620, in _env_runner
    sample_collector=sample_collector,
  File "/opt/miniconda/lib/python3.7/site-packages/ray/rllib/evaluation/sampler.py", line 1198, in _process_observations_w_trajectory_view_api
    new_episode.length - 1, filtered_obs)
  File "/opt/miniconda/lib/python3.7/site-packages/ray/rllib/evaluation/collectors/simple_list_collector.py", line 487, in add_init_obs
    assert agent_key not in self.agent_collectors
AssertionError
```

Python version: 3.7.9
OS: ubuntu 18.04
Tensorflow version: 2.4.1
Docker: Docker version 20.10.3, build 48d30b5

Logs:
[error.txt](https://github.com/ray-project/ray/files/6311071/error.txt)
[70_driver_log.txt](https://github.com/ray-project/ray/files/6311073/70_driver_log.txt)

I suspect that the error is caused by the following code:

```python
# episode.py

class MultiAgentEpisode:
   #...

    def __init__(self, policies: Dict[PolicyID, Policy],
                 policy_mapping_fn: Callable[[AgentID], PolicyID],
                 batch_builder_factory: Callable[
                     [], "MultiAgentSampleBatchBuilder"],
                 extra_batch_callback: Callable[[SampleBatchType], None],
                 env_id: EnvID):
        #...
        self.episode_id: int = random.randrange(2e9)
        #...
```

I'm using a training batch that is generating ~2k episodes/iteration. Assuming that episode ids are independent per training iteration, the job has about (if I'm not wrong in my calculations) ~20% probability of generating at least one conflicting id in first 100 training iterations (due to the [Birthday paradox](https://en.wikipedia.org/wiki/Birthday_problem#Probability_table)).

### Reproduction
I was able to reproduce the error with a the following script:

```pyhton
import ray
from ray import tune

def stop(trial_id, result):
    return result['training_iteration'] >= 1000


DEFAULT_RAY_ADDRESS = 'localhost:6379'

if __name__ == '__main__':
    horizon = 20
    num_workers = 5
    num_envs_per_worker = 128
    trainer_cpus = 11
    eval_workers = 0
    num_eval_ep = 0
    num_episodes = 5 * 3 * num_envs_per_worker * horizon
    train_batch_size = num_episodes * horizon
    sgd_minibatch_size = 1024
    num_sgd_iter = 1
    ray.init(address=DEFAULT_RAY_ADDRESS)

    tune.run(
        run_or_experiment='PPO',
        config={
            # Training settings
            'env': 'Pendulum-v0',
            'model': {
                "fcnet_hiddens": [16]
            },
            'env_config': {},
            'num_workers': num_workers,
            'num_cpus_per_worker': 1,
            'num_envs_per_worker': num_envs_per_worker,
            'rollout_fragment_length': horizon,
            'framework': 'tf2',

            # Continuous Task settings
            'horizon': horizon,
            'soft_horizon': False,
            'no_done_at_end': True,

            # Parallel Training CPU
            'num_cpus_for_driver': trainer_cpus,
            'tf_session_args': {
                'intra_op_parallelism_threads': 0,
                'inter_op_parallelism_threads': 0,
                'log_device_placement': False,
                'device_count': {
                    'CPU': trainer_cpus,
                },
                'allow_soft_placement': True,
            },
            'local_tf_session_args': {
                'intra_op_parallelism_threads': 0,
                'inter_op_parallelism_threads': 0,
            },

            # PPO specific
            'train_batch_size': train_batch_size,
            'sgd_minibatch_size': sgd_minibatch_size,
            'num_sgd_iter': num_sgd_iter,
        },
        stop=stop,
        local_dir='./logs')
```

- [X] I have verified my script runs in a clean environment and reproduces the issue.
- [X] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by andraxin**
## Why are these changes needed?

The current [circa](https://github.com/ray-project/ray/blob/master/rllib/evaluation/episode.py#L90) [31-bit](https://github.com/ray-project/ray/blob/master/rllib/evaluation/episode.py#L124) of randomness for generating `episode_id`
fails to guarantee the required [uniqueness property](https://github.com/ray-project/ray/blob/master/rllib/evaluation/episode.py#L38)
>  episode_id: Unique id identifying this trajectory.

As the _sample run_ of the enclosed _test script_ shows, collisions can occur
with as low as a few **hundred** episodes.  This impacts Ray's robustness,
resulting in a difficult to debug **random**[ assertion error](https://github.com/ray-project/ray/blob/master/rllib/evaluation/collectors/simple_list_collector.py#L642).

<details><summary>test script</summary>

```python
#!/usr/bin/env python

import random

def collision(size=2e9, run=0):
    seen = set()
    count = 0

    try:
        while True:
            eid = random.randrange(size)
            count += 1
            if eid in seen:
                break
            seen.add(eid)
    except KeyboardInterrupt:
        print(run+1, count)

    return count


if __name__ == '__main__':
    import argparse
    import pandas as pd

    pd.options.display.float_format = '{: .2f}'.format

    parser = argparse.ArgumentParser('Episode ID collision test')
    parser.add_argument('-n', '--runs', type=int, default=1000)
    parser.add_argument('-s', '--size', type=float, default=2e9)
    args = parser.parse_args()

    data = [collision(args.size, i) for i in range(args.runs)]

    print(pd.DataFrame(data).describe())
```
</details>

<details><summary>sample run</summary>

```console
$ python /tmp/test_randrange.py
               0
count    1000.00
mean    57555.97
std     31199.46
min       596.00
25%     33836.00
50%     53202.50
75%     77451.00
max    191959.00
```
</details>

This PR proposes to approximately double the amount of randomness
to circa 63 bits, by simply changing the current `2e9` limit to `1e19`.

Given that the original issue was filed in April, 2021 **and** its root cause
[identified](https://github.com/ray-project/ray/issues/15297#issuecomment-860413103) almost exactly a year ago, I'm wondering what prevented
the seemingly trivial fix.   Are there performance considerations at play?

Note that this is still a far cry from the 128 bits used when generating  
[episode IDs for `ExternalEnv`](https://github.com/ray-project/ray/blob/master/rllib/env/external_env.py#L109) (`uuid` uses 16 bytes from `os.urandom`).

## Related issue number

Closes #15297

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by pcmoritz** [comment]
Anybody looking at this @gjoliver ?

**Comment by gjoliver** [comment]
sorry I completely missed this thread.

I am thinking if we can create a scheme to make this id truly unique.
like lower 10 bits of worker_index, plus 21 bits of incremental counter.
we will then only have collision if there are more than 1K rollout workers, and 2M episodes per worker.

the number of bits we reserve for the worker_index can also be dynamic. so if you are only using 10 workers, we can give more bits to the counter.

I wonder if I am missing something, and the situation is more complicated than I think?

**Comment by andraxin** [comment]
@gjoliver 
I guess you aren't asking me, but from a quick glance
https://github.com/ray-project/ray/blob/f874053fcff02020a3312c5022b3d5cfd00e733b/rllib/evaluation/episode.py#L68
worker information may or may not be available when
this id is generated...

**Comment by andraxin** [comment]
@gjoliver @sven1977 

Before this PR gets closed as stale, I'd like to understand why it's not getting much attention.
The issue is real.  Simulations of sufficient scale are almost guaranteed to experience random
failures due to this.  The fix is simple.  What's the catch?

Are you still considering an even more unique and robust approach?

If that's going to take longer than another month or two, maybe this simple fix
could be merged as a gap filler, until you come up with a better solution....

**Review by krfricke — APPROVED** [review]
Thanks @andraxin for the PR and sorry for the delay in attending to it.

I've just merged latest master and I'm happy to merge this once tests are passing, assuming no objections from @gjoliver.

I think this could be a quick fix to alleviate problems discussed in #15297, though we may want to keep it open or track a new issue to come up with truly unique random ids.

**Comment by krfricke** [comment]
cc @sven1977 can you take a look at the failing tests and see if/how they are related?

**Comment by ArturNiederfahrenhorst** [comment]
@krfricke Failed tests are unrelated. I tried to merge master again but can't push to the PR branch.

File: rllib/evaluation/episode.py [hunk 1]
```diff
@@ -124,7 +124,8 @@ def soft_reset(self) -> None:
         same env (i.e., if `soft_horizon` is set).
         """
         self.length = 0
-        self.episode_id = random.randrange(1e19)
+        self.episode_id = random.randrange(int(1e18))
+        self.episode_id = random.randrange(int(2e9))
```
Comment (krfricke) [inline rllib/evaluation/episode.py:128]:
```suggestion
        self.episode_id = random.randrange(int(1e18))
```
?

Comment (ArturNiederfahrenhorst) [reply rllib/evaluation/episode.py]: Indeed, thanks.

**Review by ArturNiederfahrenhorst — APPROVED** [review]

**Pull request merged**